packages/test/test-utils/src/testObjectProvider.ts

```diff
@@ -122,14 +122,63 @@ function getDocumentIdStrategy(type?: TestDriverTypes): IDocumentIdStrategy {
     }
 }
 
+export class ErrorTrackingLogger extends TelemetryLogger{
+    constructor(private readonly baseLogger: ITelemetryBaseLogger){
+        super();
+    }
+
+    private readonly expectedEvents: (ITelemetryGenericEvent | undefined)[] = []
+    private readonly unexpectedErrors: ITelemetryBaseEvent[] =[];
+
+    public registerExpectedEvent(... events: ITelemetryGenericEvent[]){
+        this.expectedEvents.push(... events);
+    }
+
+    send(event: ITelemetryBaseEvent): void {
+        const ee = this.expectedEvents[0];
+        if(ee?.eventName === event.eventName){
+            let matches = true;
+            for(const key of Object.keys(ee)){
+                if(ee[key] !== event[key]){
+                    matches = false;
+                    break;
+                }
+            }
+            if(matches){
+                this.expectedEvents.shift();
+                if(event.category === "error"){
```

Comment:
I know you're hesitant to write comments, but this if block really needs it - it's the crux of what this whole thing is doing and really needs the context/purpose explained.

Reply:
happy to comment on things like this, which are potentially non-intuitive code blocks who intention isn't discoverable via the code itselft